### PR TITLE
:sparkles: Enable (longer context) granite support

### PR DESCRIPTION
**Pull request by joerunde**
## Description

Enables the upstream granite tests against `ibm-ai-platform/micro-g3.3-8b-instruct-1b` (single-card `test_granite.py::test_models` and TP=2 `test_basic_correctness.py::test_models_distributed`) by fixing a `RecursionError` in the attention backend that fired on longer prefills / larger batches. Also brings forward the platform `device_count` override and a small testing-plugin enhancement that were needed to make those tests actually run.

## Related Issues

`torch-spyre#220` / `torch-spyre#1371-3` — symbolic-offset support for `spyre.overwrite`; would let us revert to an on-device scatter and remove the CPU staging buffer added here.

## Changes

**Rewrite attention scatter-write** (`spyre_inference/v1/attention/backends/spyre_attn.py`)
- Per-token scatter into `output` had no working primitive on Spyre dim=0: `output[i:j] = ...` / `narrow().copy_(...)` silently write everything to row 0, and `torch.ops.spyre.overwrite` is deprecated. Its `compile_once` wrapper compiles one SDSC binary per unique `(shape, offset)` under `specialize_int=True`; once vLLM's model compile has filled the dynamo cache the kernel falls back to the same `compile_once` wrapper and recurses to Python's stack limit. Granite hit this in routine prefill; llama happened to compile-cache around it.
- `_online_softmax_attention` now builds the full per-attention-layer result in a CPU `output_cpu` tensor and copies it back to the Spyre output buffer in a single `output.copy_(...)` at the end of the per-sequence loop.
- Considered (and rejected) bumping `accumulated_cache_size_limit` to keep the on-device scatter: it works for short tests but compiles `N` binaries for a `query_len=N` prefill, which is unworkable as the on-device KV cache cap is raised (128K context = 128K binaries). When `torch-spyre` lands symbolic-offset overwrite, revisit and move the final write back on-device.

**Platform `device_count()` override** (`spyre_inference/platform.py`)
- `CpuPlatform.device_count()` returns 1 (CPU = single device); upstream gates like `@multi_gpu_test(num_gpus=2)` skipped on multi-card hosts as a result. Override to return `torch.spyre.device_count()` so TP-aware upstream tests actually run.

**Plugin support for tuple-parametrize override** (`tests/plugin/spyre_testing_plugin/pytest_plugin.py`)
- Some upstream tests use a single composite parametrize key like `@pytest.mark.parametrize("model, distributed_executor_backend, attention_backend, test_suite, extra_env", [...])`. The plugin's `override` hook only matched single-name keys, so the YAML couldn't redirect those tests at our model. Extends the matcher to recognise comma-joined keys (normalised on both sides) and to emit tuple values when the key is composite.

**Enable granite upstream tests** (`tests/plugin/spyre_testing_plugin/upstream_tests.yaml`)
- `test_granite.py::test_models`: `mandatory_pass` on `micro-g3.3-8b-instruct-1b` at fp16, `max_tokens=2`. Was a `model: []` placeholder waiting on the recursion fix.
- `test_basic_correctness.py::test_models_distributed`: `mandatory_pass` on `micro-g3.3-8b-instruct-1b` × `mp` executor × `CUSTOM` attention backend × test_suite `L4`, with `enable_prompt_embeds=[true]` skipped (prompt embeddings not yet supported on Spyre).
- `test_basic_correctness.py::test_models` remains skipped — its hardcoded ~5042-token prompt does not fit `MAX_MODEL_LEN_CAP=128`; orthogonal to this PR.
- No GHA workflow change needed: the existing `Upstream distributed vLLM tests` matrix entry (`-m "upstream and distributed"` on `spyre_pf_x2__pvc_1tb_x1`) already covers the newly-enabled TP=2 test, and `Upstream model vLLM tests` covers the granite single-card test.

**Doc correction** (`CLAUDE.md`)
- The "Spyre-Specific Constraints" list claimed "No tensor parallelism: Custom linear layers assume TP=1", which contradicts the in-tree `SpyreCommunicator` / `all_reduce` implementation documented at `docs/architecture/index.md` "Distributed (TP)". Corrected to "TP≥1 supported; DP>1 rejected in `TorchSpyrePlatform.check_and_update_config`".

## Test Plan

Local verification on a 2-Spyre host:

- [x] `uv run pytest -m "upstream and granite"` → 1 passed, 2172 deselected in 113s
- [x] `uv run pytest -m "upstream and distributed and model"` → 1 passed, 1 skipped (prompt-embeds), 2171 deselected in 129s
- [x] `uv run pytest -m "not upstream" tests/test_spyre_attn.py` → 73 passed, 24 skipped in 503s (no regression on the attention matrix from the CPU-buffer rewrite)
- [x] `bash format.sh` → ruff, ruff format, typos, markdownlint, GHA workflow lint, ty check all pass

## Checklist

- [X] I have read the [contributing guidelines](https://torch-spyre.github.io/spyre-inference/contributing/)
- [X] My code follows the project's code style (run `bash format.sh`)
- [X] I have added tests for my changes (if applicable) — covered by the newly-enabled upstream tests
- [X] I have updated the documentation (if applicable) — CLAUDE.md TP claim corrected
- [X] My commits include a `Signed-off-by:` line (DCO compliance)

### CODE REVIEW

File: spyre_inference/v1/attention/backends/spyre_attn.py [hunk 1]
```diff
@@ -766,6 +767,20 @@ def _online_softmax_attention(
...
+        # compiles one SDSC binary per unique offset, which recurses past the
+        # dynamo cache limit once vLLM's model compile has filled it. Raising
+        # the limit unblocks short tests but compiles N binaries for a
+        # query_len=N prefill, which doesn't scale to long contexts. Stage the
+        # result on CPU and bulk-copy at the end of the per-sequence loop.
+        # Revisit when torch-spyre lands symbolic-offset overwrite
+        # (torch-spyre#220 / #1371-3).
+        output_cpu = torch.zeros(
+            num_actual_tokens, num_heads, head_size, dtype=output.dtype, device="cpu"
+        )
```
Comment (bringlein) [inline spyre_inference/v1/attention/backends/spyre_attn.py]:
```suggestion
        output_cpu = torch.zeros_like(output, device="cpu")
```

Just small comment: This version may be more robust to layout changes? Otherwise, I would not expect that `output` shape changes.

**Review by bringlein — APPROVED** [review]
LGTM, thanks for digging into the recursion of  `_overwrite`. 

Yes, `torch.ops.spyre.overwrite` is deprecated, but we had the hope using it saves us unnecessary device-host-device cycles. However, since slicing is not supported yet on the device, they are necessary anyhow for the time being.
